Plugins/CoreAPI/server/graphql/resolvers/mutations.js

```diff
@@ -0,0 +1,435 @@
+          _.each(host.hostnames, (hostname) => {
+            if (existingHostHostnames.includes(hostname)) {
+              hadHostname = true;
+            }
+          });
+          if (hadHostname) {
+            toUpdate.push(host);
+          } else {
+            toInsert.push(host);
+          }
+        }
+      });
+    }
+
+    // if we have net-new add them...
+    if (toInsert.length > 0) {
+      _.each(toInsert, (host) => {
+        host.projectID = args.projectID;
+      });
+      let res = await PenPal.DataStore.insertMany("CoreAPI", "Hosts", toInsert);
+      _.each(res.insertedIds, (k, v) => {
+        updatedRecords.push(k);
+      });
+    }
+
+    // if we have ones to update update them....
+    _.each(toUpdate, (host) => {
+      // this is the host to add
+      // and this is the existing db entry
+      let storedHost = {};
+      let foundHost = false;
+      _.each(existingRecords, (existingHost) => {
+        if (!foundHost) {
+          if (existingHost.ipv4 === host.ipv4) {
+            storedHost = existingHost;
+            foundHost = true;
+          } else if (existingHost.mac === host.mac) {
+            storedHost = existingHost;
+            foundHost = true;
+          } else if (exhistingHost.hostnames) {
+            if (
+              existingHost.hostnames
+                .join(",")
+                .includes(host.hostnames.join(","))
+            ) {
+              storedHost = existingHost;
+              foundHost = true;
+            }
+          }
+        }
+      });
+      // at this point we have the host data to add and the existing host data.... need to diff the objects to figure out what fields need to be set....
+      // ok so merge will work but make it so that we can't remove fields... need to find out way to intelligently find the difference and adjust accordingly...
+      let storedHostnames = [];
+      if (storedHost.hostnames) {
+        storedHostnames = [].concat(storedHost.hostnames);
+      }
+      let mergedObject = _.merge(storedHost, host);
+      let ID = mergedObject._id;
+      delete mergedObject._id;
+      _.each(storedHostnames, (hostname) => {
+        if (
+          mergedObject.hostnames &&
+          !mergedObject.hostnames.includes(hostname)
+        ) {
+          mergedObject.hostnames.push(hostname);
+        }
+      });
+      // After we merge but before we insert we need to see if any splitout attributes were added.... OS is a single embedded object
+      // so that's ok but hosts also have files, notes, and services....
+      if (mergedObject.services) {
+        console.log("HAVE SERVICES TO HANDLE....");
+        delete mergedObject.services;
+      }
+      mergedObject.projectID = args.projectID;
+      let res = PenPal.DataStore.update(
+        "CoreAPI",
+        "Hosts",
+        { _id: ID },
+        mergedObject
+      );
+      if (res > 0) updatedRecords.push(ID);
+    });
+
+    // TODO - Transactions?
+    return {
```

Comment:
There's a lot of logic in here -- I trust that you tested adequately, but we should maybe consider adding automated unit testing before going "live"

Reply:
sounds good, will make note to add unit tests and will add once quick mvp is figured out